extension/js/common/api/email-provider/gmail/google-auth-helper.ts

```diff
@@ -0,0 +1,223 @@
+    const toSave: AcctStoreDict = {
+      full_name: full_name || parsedOpenId.name, // eslint-disable-line @typescript-eslint/naming-convention
+      picture: picture || parsedOpenId.picture,
+    };
+    if (typeof tokensObj.refresh_token !== 'undefined') {
+      toSave.google_token_refresh = tokensObj.refresh_token;
+    }
+    await AcctStore.set(acctEmail, toSave);
+    await InMemoryStore.set(acctEmail, InMemoryStoreKeys.ID_TOKEN, tokensObj.id_token);
+    await InMemoryStore.set(acctEmail, InMemoryStoreKeys.GOOGLE_TOKEN_ACCESS, tokensObj.access_token, googleTokenExpires);
+  };
+
+  public static apiGoogleAuthCodeUrl = (authReq: AuthReq) => {
+    /* eslint-disable @typescript-eslint/naming-convention */
+    return Url.create(GoogleAuthHelper.OAUTH.url_code, {
+      client_id: GoogleAuthHelper.OAUTH.client_id,
+      response_type: 'code',
+      access_type: 'offline',
+      prompt: 'consent',
+      state: authReq.expectedState,
+      redirect_uri: GoogleAuthHelper.OAUTH.redirect_uri,
+      scope: (authReq.scopes || []).join(' '),
+      login_hint: authReq.acctEmail,
+    });
+    /* eslint-enable @typescript-eslint/naming-convention */
+  };
+
+  public static googleAuthRefreshToken = async (refreshToken: string) => {
+    return (await GoogleAuthHelper.ajax(
+      {
+        /* eslint-disable @typescript-eslint/naming-convention */
+        url: Url.create(GoogleAuthHelper.OAUTH.url_tokens, {
+          grant_type: 'refresh_token',
+          refreshToken,
+          client_id: GoogleAuthHelper.OAUTH.client_id,
+          client_secret: GoogleAuthHelper.OAUTH.client_secret,
+        }),
+        /* eslint-enable @typescript-eslint/naming-convention */
+        method: 'POST',
+        crossDomain: true,
+        async: true,
+      },
+      Catch.stackTrace()
+    )) as unknown as GoogleAuthTokensResponse;
+  };
+
+  public static googleAuthGetTokens = async (code: string) => {
+    return (await GoogleAuthHelper.ajax(
+      {
+        /* eslint-disable @typescript-eslint/naming-convention */
+        url: Url.create(GoogleAuthHelper.OAUTH.url_tokens, {
+          grant_type: 'authorization_code',
+          code,
+          client_id: GoogleAuthHelper.OAUTH.client_id,
+          client_secret: GoogleAuthHelper.OAUTH.client_secret,
+          redirect_uri: GoogleAuthHelper.OAUTH.redirect_uri,
+        }),
+        /* eslint-enable @typescript-eslint/naming-convention */
+        method: 'POST',
+        crossDomain: true,
+        async: true,
+      },
+      Catch.stackTrace()
+    )) as unknown as GoogleAuthTokensResponse;
+  };
+
+  /// Duplicated code begin
+  // This logic was duplicated (It's exactly same as Api.ajax) but added here because we can't use Api.ajax here as it will cause circular dependency warning
+  private static ajax = async (req: JQueryAjaxSettings, stack: string): Promise<unknown | JQuery.jqXHR<unknown>> => {
```

Comment:
Let me know your thought about this block if you agree on duplicating logi or not.

Reply:
Sounds good.